### DISCUSSION TITLE
Strain at t_points at shell integration points

### DISCUSSION THREAD

**Discussion by Zoophish** [idaholab/moose#22460] [Q&A Modules: Solid mechanics] [answered]
Hi, I'm trying to extract curvature values from shell elements by using some AuxVariables. To do this, I'm accessing the strain values at the two t_points above and below the shell mid surface at each quadrature point using the `RankTwoTensor` AuxKernel, like in the [modules/tensor_mechanics/test/tests/shell/static/straintest.i](https://mooseframework.inl.gov/source/materials/ADComputeIncrementalShellStrain.html#6fc19356-71b7-4c16-a433-99ac65691c3b) example.

If I use the material property `total_global_strain_t_points_`, like in the example, the strain values all output zero at every element, despite my model being under load and producing non-zero displacement. Interestingly, the `global_stress_t_points_` material property outputs valid stress values, so it seems to only be an issue with the strain.

Does anyone know the correct way to extract the strain values at the t_points, or why they may be outputting zero? I will add my input file as a comment.

Thanks,

Sam

**Comment by Zoophish** [comment]
This is my input file:

```
x_size = 100 #mm
y_size = 25 #mm
n_plies = 16
ply_thickness = '${fparse 5/n_plies}'
n_elems_x = 64
# n_elems_y should be even to allow for the creation of the centerline sideset
n_elems_y = 16
force = 500 # N
pressure = '${fparse force / (y_size)}'

[Mesh]
  [Generator]
    type = GeneratedMeshGenerator
    dim = 2
    xmax = ${x_size}
    ymax = ${y_size}
    nx = ${n_elems_x}
    ny = ${n_elems_y}
  []
  [SampleXCenterline]
    type = ParsedGenerateSideset
    input = Generator
    new_sideset_name = Centerline
    combinatorial_geometry = 'x=${fparse x_size/2}'
  []
[]

[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [disp_z]
  []
  [rot_x]
  []
  [rot_y]
  []
[]

[AuxVariables]
  [strain_xx_bottom]
    order = CONSTANT
    family = MONOMIAL
  []
  [strain_xx_top]
    order = CONSTANT
    family = MONOMIAL
  []
  [strain_yy_bottom]
    order = CONSTANT
    family = MONOMIAL
  []
  [strain_yy_top]
    order = CONSTANT
    family = MONOMIAL
  []
  [strain_xx_midplane]
    order = CONSTANT
    family = MONOMIAL
  []
  [kxx]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [strain_xx_bottom]
    type = RankTwoAux
    variable = strain_xx_bottom
    # quadrature underneath the midplane
    rank_two_tensor = total_global_strain_t_points_0
    index_i = 0
    index_j = 0
  []
  [stress_xx_top]
    type = RankTwoAux
    variable = strain_xx_top
    # quadrature above the midplane
    rank_two_tensor = global_stress_t_points_1
    index_i = 0
    index_j = 0
  []
  [strain_yy_bottom]
    type = RankTwoAux
    variable = strain_yy_bottom
    # quadrature underneath the midplane
    rank_two_tensor = total_global_strain_t_points_0
    index_i = 1
    index_j = 1
  []
  [strain_yy_top]
    type = RankTwoAux
    variable = strain_yy_top
    # quadrature above the midplane
    rank_two_tensor = total_global_strain_t_points_1
    index_i = 1
    index_j = 1
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  rotations = 'rot_x rot_y'
[]

[Kernels]
  [solid_disp_x]
    type = ADStressDivergenceShell
    block = '0'
    component = 0
    variable = disp_x
    through_thickness_order = SECOND
  []
  [solid_disp_y]
    type = ADStressDivergenceShell
    block = '0'
    component = 1
    variable = disp_y
    through_thickness_order = SECOND
  []
  [solid_disp_z]
    type = ADStressDivergenceShell
    block = '0'
    component = 2
    variable = disp_z
    through_thickness_order = SECOND
  []
  [solid_rot_x]
    type = ADStressDivergenceShell
    block = '0'
    component = 3
    variable = rot_x
    through_thickness_order = SECOND
  []
  [solid_rot_y]
    type = ADStressDivergenceShell
    block = '0'
    component = 4
    variable = rot_y
    through_thickness_order = SECOND
  []
[]

[Materials]
  # Materials are computed using the shell functions, see the moose docs on Shells.
  [stress]
    type = ADComputeShellStress
    through_thickness_order = SECOND
  []
  [elasticity]
    # isotropic aluminium
    type = ADComputeIsotropicElasticityTensorShell
    youngs_modulus = 70e3
    poissons_ratio = 0.31
    through_thickness_order = SECOND
    block = 0
  []
  [strain]
    type = ADComputeFiniteShellStrain
    thickness = '${fparse ply_thickness*n_plies}'
    through_thickness_order = SECOND
  []
[]

[BCs]
  [Fixed_X]
    type = ADDirichletBC
    boundary = left
    value = 0
    variable = disp_x
  []
  [Fixed_Y]
    type = ADDirichletBC
    boundary = Centerline
    value = 0
    variable = disp_y
  []
  [Fixed_Z]
    type = ADDirichletBC
    boundary = 'left right'
    value = 0
    variable = disp_z
  []
  [Load]
    type = ADNeumannBC
    boundary = Centerline
    value = '${fparse -pressure/2}'
    variable = disp_z
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = false
  []
[]

[Executioner]
  type = Transient
  dt = 1
  dtmin = 1
  end_time = 1
  automatic_scaling = TRUE
  solve_type = 'NEWTON'
  line_search = 'default'
  petsc_options_iname = '-pc_type  -snes_linesearch_type -pc_factor_shift_type -snes_max_it  -ksp_max_it'
  petsc_options_value = 'lu        basic                 NONZERO               20            30'
[]

[Postprocessors]
  [Strain_XX]
    type = ElementalVariableValue
    elementid = 543
    variable = strain_xx_bottom
  []
[]

[Outputs]
  [output_file]
    type = Exodus
    file_base = shell_cantilever_stress_gradient
  []
[]
```

**Comment by GiudGiud** [comment]
@neuphris @recuero

**Comment by Zoophish** [comment, marked as answer]
So I just discovered that changing `ADComputeFiniteShellStrain` to `ADComputeIncrementalShellStrain` solves the issue, however I am not sure why this is. Would anyone be able to explain?

On a slight side note, I was also having an issue using `ADComputeIncrementalShellStrain` where it was giving the error "Shell element needs to have exactly four quadrature points." for the above input file. However, I verified that all of the elements in the model had exactly 4 quadrature points using a custom AuxKernel. I removed the error-throwing code in the `ADComputeIncrementalShellStrain` C++ file,  recompiled it and it worked. Could this be a bug?

**Reply by neuphris** [reply]
My first guess that " Shell Element needs to have exactly four quadrature points"   may have been generated due to the BC that was developed to be applied at the area and is being applied at the edge (I don't know how ADNeuman BC works) because when I tried to create a test with pressure BC (that should be applied over the area) at the edge of shell element, it led to the failure of the test cases and we decided to generate that error for such mismatch. Refer to the comment at the end of https://github.com/idaholab/moose/discussions/20771

**Reply by Zoophish** [reply]
Okay that's interesting, I think my workaround for now will be to comment out the error as I seem to be getting valid results with the Neumann boundary.

Would you be able to explain why `ADComputeFiniteShellStrain` does not work?

**Reply by neuphris** [reply]
I went through the code quickly and found that the strains in the contravariant system are not converted to the global strain values which would have been stored in _total_global_strain variable for output in the ADComputeFiniteShellStrain due to which you are getting zero values for the strain.

The stress are correct because the strains in contravariant system are used to calculate the stress in the covariant  system and is converted into the global system. 

I think it should be an easy fix. Just adding the few lines of code in theend of  ADFiniteShellStrain::computeProperties() (similar to ADIncrementalShellStrain::computeProperties())

      for (unsigned int ii = 0; ii < 3; ++ii)
        for (unsigned int jj = 0; jj < 3; ++jj)
          _unrotated_total_strain(ii, jj) = MetaPhysicL::raw_value((*_total_strain[j])[i](ii, jj));
      (*_total_global_strain[j])[i] = (*_contravariant_transformation_matrix[j])[i] *
                                      _unrotated_total_strain *
                                      (*_contravariant_transformation_matrix[j])[i].transpose();

**Reply by Zoophish** [reply]
Thanks for the help, I'll try out those changes.

**Reply by neuphris** [reply]
@bwspenc We never noticed this. It needs to be fixed (i haven't tried, but it seems like the above-mentioned change is the only thing missing) and merged.

**Reply by Zoophish** [reply]
I can confirm the changes you suggested worked when I added it to my copy of MOOSE and recompiled.
